### PR TITLE
Lower firrtl::memop to generator modules

### PR DESCRIPTION
**Pull request by darthscsi**
Don't try to implement memories during lowering, encapsulate them for dealing with later.

### CODE REVIEW

**Comment by darthscsi** [comment]
- [ ] update testcases
- [ ] merge main

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 1]
```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
...
+#undef cmp3way
+  }
+};
+} // namespace
+
+static std::string getFirMemoryName(const FirMemory &mem) {
+  return llvm::formatv("FIRRTLMem_{0}_{1}_{2}_{3}_{4}_{5}_{6}",
+                       mem.numReadPorts, mem.numWritePorts,
+                       mem.numReadWritePorts, mem.dataWidth, mem.depth,
+                       mem.readLatency, mem.writeLatency);
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:147]: Does it make sense to add some indicator of what these are, e.g.:

```
"FIRRTLMem_r{0}w{1}rw{2}_{3}x{4}_rlat{5}_wlat{6}"
```

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 2]
```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
...
+    else if (portKind == MemOp::PortKind::Write)
+      ++numReadPorts;
+    else
+      ++numReadWritePorts;
+  }
+  auto width = op.getDataType().getBitWidthOrSentinel();
+  if (width <= 0) {
+    op.emitError("'firrtl.mem' should have simple type and known width");
+    width = 0;
+  }
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:168]: Is this check actually looking at `<= 0`, i.e., "memory has unknown width (`-1`) or zero width"? What does "simple type" mean?

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 3]
```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
...
+          op.depth(),   op.readLatency(), op.writeLatency()};
+}
+
+static std::set<FirMemory> collectFIRRTLMemories(Operation *op) {
+  auto module = cast<FModuleOp>(op);
+  std::set<FirMemory> retval;
+  for (auto op : module.getBody().getOps<MemOp>())
+    retval.insert(analyzeMemOp(op));
+  return retval;
+}
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: What is the motivation for using `std::set` vs. `SmallSet` or something from the LLVM collections? I think you just want stable iteration order, yes? 

A vector, sorted and uniqued would seem to work here (one of the alternatives the LLVM Programmer's Manual suggests).

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 4]
```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
...
+
+static std::set<FirMemory> mergeFIRRTLMemories(std::set<FirMemory> a,
+                                               std::set<FirMemory> b) {
+  a.insert(b.begin(), b.end());
+  return a;
+}
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: Is this returning a copy of `a`? Can this be done more efficiently, e.g., returning a reference to the mutated `a`?

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 5]
```diff
@@ -243,6 +332,72 @@ void FIRRTLModuleLowering::runOnOperation() {
...
 }
 
+void FIRRTLModuleLowering::lowerMemoryDecls(const std::set<FirMemory> &mems,
+                                            Block *top,
+                                            CircuitLoweringState &state) {
+  auto b =
+      ImplicitLocOpBuilder::atBlockBegin(UnknownLoc::get(&getContext()), top);
+  std::array<StringRef, 6> schemaFields = {"depth",         "numReadPorts",
+                                           "numWritePorts", "readLatency",
+                                           "writeLatency",  "width"};
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:356]: Consider `SmallVector<StringRef, 6>`.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 6]
```diff
@@ -243,6 +332,72 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+
+  SmallVector<rtl::ModulePortInfo> ports;
+  size_t inputPin = 0;
+  size_t outputPin = 0;
+
+  Type b1Type = IntegerType::get(&getContext(), 1);
+
+  auto makePortCommon = [&](StringRef prefix, Twine i, Type bAddrType) {
+    ports.push_back({b.getStringAttr((prefix + Twine("_clock_") + i).str()),
+                     rtl::INPUT, b1Type, inputPin++});
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:370]: The `Twine("_clock_")` may be able to be simplified to just `"_clock_"`.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 7]
```diff
@@ -243,6 +332,72 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+    ports.push_back({b.getStringAttr((prefix + Twine("_en_") + i).str()),
+                     rtl::INPUT, b1Type, inputPin++});
+    ports.push_back({b.getStringAttr((prefix + Twine("_addr_") + i).str()),
+                     rtl::INPUT, bAddrType, inputPin++});
+  };
+
+  for (auto &mem : mems) {
+    Type bDataType;
+    if (mem.dataWidth > 0)
+      bDataType = IntegerType::get(&getContext(), mem.dataWidth);
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:380]: This doesn't handle the `mem.dataWidth == 0` case. Is that okay, e.g., have we fully removed zero-width memories before this?

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 8]
```diff
@@ -243,6 +332,72 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+      ports.push_back({b.getStringAttr((Twine("wr_mask_") + Twine(i)).str()),
+                       rtl::INPUT, bDataType, inputPin++});
+      if (mem.dataWidth > 0)
+        ports.push_back({b.getStringAttr((Twine("wr_data_") + Twine(i)).str()),
+                         rtl::INPUT, bDataType, inputPin++});
+    }
+    assert(mem.numReadWritePorts == 0 &&
+           "Read/Write ports should be eliminated");
+    std::array<NamedAttribute, 6> genAttrs = {
+        b.getNamedAttr("depth", b.getUI32IntegerAttr(mem.depth)),
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:400]: `mem.depth` is 64-bits in the ODS definition.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 9]
```diff
@@ -1521,258 +1689,84 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
...
+        portKind == MemOp::PortKind::Read
+            ? readCount
+            : portKind == MemOp::PortKind::Write ? writeCount : readwriteCount;
+    auto addInput = [&](SmallVectorImpl<Value> &operands, StringRef field,
+                        size_t width) {
+      auto portType = IntegerType::get(op.getContext(), width);
+      auto accesses = getAllFieldAccesses(op.getResult(i), field);
+
+      Value wire = createTmpWireOp(
+          portType, ("." + portName + "." + field + ".wire").str());
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1733]: Should this use `.` or `_`? I thought we had some problems with `.` printing as `U23` or something?

**Review by seldridge — COMMENTED** [review]
This is coming together great. I love the negative line count and removal of my janky code. 😬 

Some questions throughout, but nothing critical. The output tests look good.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 5 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:356]: C-style array (std::array) is more applicable here, as there is no reason to imply dynamic size.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:147]: This is intended as a unique symbol name, not a meaning human readable name.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:168]: simple is non struct or vector.  LowerTypes needs to lower memories with complex data types.  #888  This will also give general write mask behavior.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: It's done though I don't know that it matters much.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:168]: But, it continues with zero width as due to the subfield accesses later, it is easier to lower degenerate memories and emit an error than to not lower them and emit an error.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 7 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:380]: It does, but IntegerType can't be zero.  The data ports are elided for zero, so we don't need the type.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 9 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1733]: I was keeping the convention of temporary wire names, for which there were some strong opinions on keeping as ".".  As I don't have strong opinions on that...

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 2 above)
Comment (seldridge) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:168]: In FIRRTL-lingo, "non-struct or vector" would be "ground type" as opposed to "aggregate type".

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 9 above)
Comment (seldridge) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1733]: Not blocking. I'm fine with it as is. :+1:

**Comment by darthscsi** [comment]
Closes #787

**Comment by mikeurbach** [comment]
So is the idea here to only go the blackbox route for memories? I'm trying to parse this comment "Don't try to implement memories during lowering, encapsulate them for dealing with later."

I just gave a talk on CIRCT and one of the points I made was that we _do_ have ways to deal with memory in a whitebox way... so if that's going away, I'm curious what the recommended approach is. Does it rely on something third-party? I haven't been keeping up with the generator modules and all that.

cc @JuanEsco063 for visibility, since I guess we may discuss these topics in half an hour.

**Comment by seldridge** [comment]
@mikeurbach: This can be handled either in a whitebox or blackbox way. The idea being that something lowered to a generator op doesn't have a single, defined way of being further lowered. E.g., it's a memory and it needs to map to the output of a memory compiler or an FPGA simulation model. Anything which lowers the generator schema can be used to replace it with a blackbox (e.g., some Verilog/VHDL blob in a file) or a whitebox (e.g., RTL Dialect).

@darthscsi: This does highlight the fact that we're removing the only way to lower FIRRTL memories (my janky memory lowering code) and not providing a replacement, yet. Does it make sense to preserve the current FIRRTL -> RTL memory lowering into a pass that can handle the FIRRTL memory schema? It only occurred to me now that this is load bearing for other dialects. 😬

**Comment by darthscsi** [comment]
We will have an in tree memory lower, though it is true it doesn't exist yet.

**Comment by mikeurbach** [comment]
Thanks for weighing in Schuyler. I think it makes sense, and I am starting to understand the generator concept, so let me try to rephrase. If the existing functionality got hatcheted out into some other pass (instead of just dropped), we could have the lowering in this PR, and then in the short term we could use the new pass to further lower to RTL dialect. Or, we could do something totally separate--it's up to us.

In terms of the fact that this is load bearing for the Handshake lowering Juan and I have been doing, yes, we do use your memory lowering code, and it has worked before for Xilinx FPGAs (under some conditions, maybe with manual tweaks, etc.). It's been brought up time and again that Handshake really doesn't need to lean so much on FIRRTL, so if this is another case where we are just doing something that's not supported, I'm fine with working out another way to get Handshake memories synthesized on FPGAs.

**Comment by mikeurbach** [comment]
Oh we posted at the same time, thanks for the further info Andrew.. I guess having an in-tree memory lowering that is blessed by the FIRRTL folks would be great for the Handshake stuff in the medium term.

**Comment by JuanEsco063** [comment]
@mikeurbach Thank you for keeping in the loop. Too bad we didn't have time today to discuss this so hopefully next week. 

The cases I am considering are agnostic to how we handle the memories generated in the RTL (could be defined as registers, BRAM, black box, etc). Of course I imagine different use cases would benefit from one representation or the other. In my case, I am more interested in knowing if there is a consensus, or at least start the discussion, on how to do the load/store operation to those internal memory elements from/to an external entity in order to integrate the generated RTL, say an GEMM accelerator, in a bigger design.

**Comment by darthscsi** [comment]
> @mikeurbach Thank you for keeping in the loop. Too bad we didn't have time today to discuss this so hopefully next week.
> 
> The cases I am considering are agnostic to how we handle the memories generated in the RTL (could be defined as registers, BRAM, black box, etc). Of course I imagine different use cases would benefit from one representation or the other. In my case, I am more interested in knowing if there is a consensus, or at least start the discussion, on how to do the load/store operation to those internal memory elements from/to an external entity in order to integrate the generated RTL, say an GEMM accelerator, in a bigger design.

For reasoning about memories, we will likely want them to be first-class entities.  This is just about taking firrtl memories and making lower-to-rtl agnostic to their implementation.  Later stages will replace the generator modules with a specific implementation.  We are doing this because we don't have first-class memories outside firrtl, I fully assume when we do, we'll use those directly.

**Comment by darthscsi** [comment]
#915 implements lowering in tree for these nodes.

**Comment by mikeurbach** [comment]
Awesome, I am checking it out! I am testing some Handshake programs that create FIRRTL memory nodes right now. Rather than debugging the old approach, I will test the new way with generator modules.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 10]
```diff
@@ -243,6 +350,86 @@ void FIRRTLModuleLowering::runOnOperation() {
   circuit.erase();
 }
 
+void FIRRTLModuleLowering::lowerMemoryDecls(const SmallVector<FirMemory> &mems,
+                                            Block *top,
+                                            CircuitLoweringState &state) {
+  auto b =
+      ImplicitLocOpBuilder::atBlockBegin(UnknownLoc::get(&getContext()), top);
+  std::array<StringRef, 7> schemaFields = {
+      "depth",        "numReadPorts", "numWritePorts", "readLatency",
```
Comment (mikeurbach) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:359]: Should "numReadWritePorts" be present in the schema?

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 11]
```diff
@@ -243,6 +350,86 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+      makePortCommon("wo", Twine(i), bAddrType);
+      ports.push_back({b.getStringAttr(("wo_mask_" + Twine(i)).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      if (mem.dataWidth > 0)
+        ports.push_back({b.getStringAttr(("wo_data_" + Twine(i)).str()),
+                         rtl::INPUT, bDataType, inputPin++});
+    }
+    std::array<NamedAttribute, 7> genAttrs = {
+        b.getNamedAttr("depth", b.getI64IntegerAttr(mem.depth)),
+        b.getNamedAttr("numReadPorts", b.getI32IntegerAttr(mem.numReadPorts)),
```
Comment (mikeurbach) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:417]: Should this be `getUI32IntegerAttr`? It's unsigned int for "numWritePorts".

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:417]: Furthermore, should there be a "numReadWritePorts" here?

**Review by mikeurbach — COMMENTED** [review]
A couple minor comments based on trying this out with #915. I may be wrong, but it seems like #915 assumes these things are there.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 11 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: did I not push?  I did not..

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 12]
```diff
@@ -243,6 +350,88 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+                       b1Type, inputPin++});
+      ports.push_back({b.getStringAttr((prefix + "_addr_" + i).str()),
+                       rtl::INPUT, bAddrType, inputPin++});
+    };
+
+    Type bDataType;
+    if (mem.dataWidth > 0)
+      bDataType = IntegerType::get(&getContext(), mem.dataWidth);
+    Type bAddrType =
+        IntegerType::get(&getContext(), llvm::Log2_64_Ceil(mem.depth));
```
Comment (mikeurbach) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:386]: `Log2_64_Ceil` is 0 when the depth is 1, which returns an `i0` type here. I think that edge case chokes #915. I haven't tracked down exactly where the mismatch is, but when using this PR and #915, I see this error:

```
'rtl.instance' op #2 operand type must be 'i0', but got 'i1'
...
note: see current operation: %mem0.ro_data_0 = "rtl.instance"(%clock, %arg2_valid, %0, %clock, %2, %1, %43, %arg0_data) {instanceName = "mem0", moduleName = @FIRRTLMem_1_1_0_32_1_0_1_1} : (i1, i1, i1, i1, i1, i1, i1, i32) -> i32
```

I know there is a PR out to canonicalize away 0-wide memories, but what's the approach for 1-wide memories? Should that be handled here? Should there be a separate pass to turn such memories into registers? Not sure what the thinking is, but there appears to be a mismatch between this PR and what #915 does.

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:386]: Confirmed this is just an edge case. When I updated my silly test to be a 2-wide memory instead of 1-wide, this pass and #915 work well together! Now to test it in the simulator...

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:386]: Works for me in the simulator.

Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: Nice.  I decided to make the minimum bus width always 1-bit (address, data), to keep the port signatures consistent.  Since FIRRTL already allows wider and narrower addresses than needed for a memory to be used, this seems well within the accepted behavior.

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: Sounds good to me.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 13]
```diff
@@ -1520,258 +1719,98 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
...
-      auto rd_lastdata = last.rd_data;
-      addToAlwaysFFBlock(sv::EventControl::AtPosEdge, clk, [&]() {
-        addIfProceduralBlock(cond, [&]() { buildPAssign(slot, rd_lastdata); });
-      });
-      continue;
-    }
+    if (portKind == MemOp::PortKind::Read) {
+      addInput(readOperands, "clk", 1);
+      addInput(readOperands, "en", 1);
+      addInput(readOperands, "addr", llvm::Log2_64_Ceil(memSummary.depth));
```
Comment (mikeurbach) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1783]: I think these uses of Log2_64_Ceil will also need the `std::max(1, ...)` treatment.

```suggestion
      addInput(readOperands, "addr", std::max(1U, llvm::Log2_64_Ceil(memSummary.depth)));
```

Can probably just compute that once before the if/else branch

Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1783]: The max was moved into add Input and should be removed there, thanks.

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1783]: Actually, I was testing with the latest revision, but there still appears to be an issue. The max is used here: https://github.com/llvm/circt/pull/864/files#diff-fb69b8a7ca0d098b8278662283e61f2dd1564898caa902d72c79b2299f833a8eR1748, but below the `if (width)` doesn't take into account the fact that width may have been bumped from 0 to 1. Is that intended? LowerToRTL was failing with an assert the way it is now, for a depth-1 memory.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 14]
```diff
@@ -1520,258 +1719,99 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
...
+    auto portName = op.getPortName(i).getValue();
+    auto portKind = *op.getPortKind(i);
+
+    auto &portKindNum =
+        portKind == MemOp::PortKind::Read
+            ? readCount
+            : portKind == MemOp::PortKind::Write ? writeCount : readwriteCount;
+    auto addInput = [&](SmallVectorImpl<Value> &operands, StringRef field,
+                        size_t width) {
+      auto portType =
```
Comment (mikeurbach) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1747]: @darthscsi I guess what I'm saying is to add something like 
```suggestion
width = std::max(1U, width);
```
Up here at the top, so all the uses of `width` have this effect. Is that correct? Seems to work locally for me.

Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1747]: Correctly handling the subfieldop requires knowing the original width.  That could be recomputed from the type or it could be passed along.  I don't see a compelling way to do it either way.

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1747]: Hmm ok I guess that makes sense. I guess there is more to this than I was understanding.

If it helps, here is a boiled down testcase for a memory with depth = 1, which should be suitable for dropping into lower-to-rtl.mlir: [test.txt](https://github.com/llvm/circt/files/6307350/test.txt)

That will hit this assertion on the current revision: https://github.com/llvm/circt/blob/260179d2123875b0ad295474fbf39f616586154f/lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp#L1358

Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1748]: Thanks, fixed.

Comment (mikeurbach) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1748]: Awesome, thanks.

**Comment by darthscsi** [comment]
Closes #844

**Comment by mikeurbach** [comment]
:rocket:

**Comment by lattner** [comment]
I finally got around to reviewing this -- it looks like there is a bunch of duplicated code between the `RTLMemSimImpl` and `LowerToRTL` pass, including duplication of the FirMemory struct and `analyzeMemOp`.  Is there a way to factor this, e.g. shove the struct into a header file somewhere, e.g. make this a method on MemOp?

**Comment by lattner** [comment]
I filed this issue to track this https://github.com/llvm/circt/issues/1033

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 15]
```diff
@@ -243,6 +350,85 @@ void FIRRTLModuleLowering::runOnOperation() {
...
+
+    for (size_t i = 0; i < mem.numWritePorts; ++i) {
+      makePortCommon("wo", Twine(i), bAddrType);
+      ports.push_back({b.getStringAttr(("wo_mask_" + Twine(i)).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr(("wo_data_" + Twine(i)).str()),
+                       rtl::INPUT, bDataType, inputPin++});
+    }
+    std::array<NamedAttribute, 8> genAttrs = {
+        b.getNamedAttr("depth", b.getI64IntegerAttr(mem.depth)),
```
Comment (lattner) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:413]: This looks like it should be a helper method on FirMemory, to return all the attrs etc.
